net_buf: Fixed race condition in net_buf_ref() / net_buf_unref() functions

dwyoung43 opened the pull request:
which could cause an incorrect buffer ref count when the same buffer was referenced/dereferenced from two different threads, causing a memory leak.

jhedberg reviewed (changes requested):
I'm not opposed to adding such a feature (yes, it's a feature - the current design is a known compromise that was acceptable when the use cases for net_buf were limited and all known).

That said, it's possible that net_buf will at some point do all this using Zephyr's atomic API, meaning `atomic_t`, `atomic_dec()` and `atomic_inc()`. I think it would therefore make sense to create local helpers which mimic those `dec()` & `inc()` APIs, to ease the transition to `atomic_t` (if this ever happens it might also be a compile-time option to let people use the more compact representation we have today).

Zephyr has an existing C-only implementation for `atomic_t` for architectures which don't offer any native implementation of this. I think you can likely just copy that implementation except the underlying type is the current `uint8_t`:
https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/atomic_c.c

dwyoung43 committed:
sys: atomic: Add atomic_uchar_t with C11 stdatomic

Introduce atomic_uchar_t as a new atomic type backed by C11 stdatomic.
Add a complete set of atomic_uchar_*() operations (set, get, cas, add,
sub, inc, dec, or, xor, and, nand, test_bit) in atomic.h, mirroring
the existing atomic_t API.

This enables atomic byte-width operations for use cases such as
reference counting where a full atomic_t is unnecessarily wide.

Signed-off-by: Doug Young <dougyoung@meta.com>

dwyoung43 committed:
net_buf: Use atomic_uchar_t for ref counting
